src/io/flutter/actions/ExtractWidgetAction.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2018 The Chromium Authors. All rights reserved.
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
+package io.flutter.actions;
+
+import com.intellij.openapi.actionSystem.AnActionEvent;
+import com.intellij.openapi.actionSystem.DataContext;
+import com.intellij.openapi.actionSystem.PlatformDataKeys;
+import com.intellij.openapi.editor.Caret;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.project.DumbAwareAction;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.ui.DocumentAdapter;
+import com.intellij.util.ui.JBUI;
+import com.jetbrains.lang.dart.ide.refactoring.ServerRefactoringDialog;
+import io.flutter.refactoring.ExtractWidgetRefactoring;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.swing.*;
+import javax.swing.event.DocumentEvent;
+import java.awt.*;
+
+public class ExtractWidgetAction extends DumbAwareAction {
+  @Override
+  public void actionPerformed(AnActionEvent event) {
+    final DataContext dataContext = event.getDataContext();
+    final Project project = dataContext.getData(PlatformDataKeys.PROJECT);
+    final VirtualFile file = dataContext.getData(PlatformDataKeys.VIRTUAL_FILE);
+    final Editor editor = dataContext.getData(PlatformDataKeys.EDITOR);
+    final Caret caret = dataContext.getData(PlatformDataKeys.CARET);
+
+    if (project != null && file != null && editor != null && caret != null) {
+      final ExtractWidgetRefactoring refactoring = new ExtractWidgetRefactoring(project, file, caret.getOffset(), 0);
+      new ExtractWidgetDialog(project, editor, refactoring).show();
+    }
+  }
+}
+
+class ExtractWidgetDialog extends ServerRefactoringDialog<ExtractWidgetRefactoring> {
+  @NotNull final ExtractWidgetRefactoring myRefactoring;
+  private final JTextField myNameField = new JTextField();
+  private final JCheckBox myStatefulCheckBox = new JCheckBox("Extract StatefulWidget");
+
+  public ExtractWidgetDialog(@NotNull Project project,
+                             @Nullable Editor editor,
+                             @NotNull ExtractWidgetRefactoring refactoring) {
+    super(project, editor, refactoring);
+    myRefactoring = refactoring;
+    setTitle("Extract Widget");
+    init();
+
+    myNameField.setText("NewWidget");
```

Comment:
We don't know whether this new name will be legal `NewWidget` - if we did we could do something like `NewWidget7`.

Another option is to autogenerate interesting names that users would need to change. So, `BluePony` or `TinySunrise`; this is what DartPad does. I'm not advocating for this, but thought I would mention it.